File: scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/kms/SimpleKeyManagementService.scala
/*
 * Copyright 2016 The BigDL Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.bigdl.ppml.kms

import com.intel.analytics.bigdl.dllib.utils.Log4Error

import scala.collection.mutable.HashMap
import scala.util.Random
import org.apache.hadoop.conf.Configuration
import com.intel.analytics.bigdl.ppml.utils.KeyReaderWriter

class SimpleKeyManagementService protected(
      simpleAPPID: String,
      simpleAPIKEY: String) extends KeyManagementService {
  val enrollMap = new HashMap[String, String]
  val keyReaderWriter = new KeyReaderWriter
  var _appId: String = _
  var _apiKey: String = _

  setAppIdAndKey(simpleAPPID, simpleAPIKEY)

  Log4Error.invalidInputError(simpleAPPID != "", s"simpleAPPID should not be empty string.")
  Log4Error.invalidInputError(simpleAPIKEY != "", s"simpleAPIKEY should not be empty string.")

  def retrievePrimaryKey(primaryKeySavePath: String, config: Configuration = null): Unit = {
    timing("SimpleKeyManagementService retrievePrimaryKey") {
      Log4Error.invalidInputError(enrollMap.keySet.contains(_appId) &&
        enrollMap(_appId) == _apiKey, "appid and apikey do not match!")
      Log4Error.invalidInputError(primaryKeySavePath != null && primaryKeySavePath != "",
        "primaryKeySavePath should be specified")
      val suffix = (1 to 4).map { x => Random.nextInt(10) }.mkString
      val encryptedPrimaryKey = _appId + suffix
      keyReaderWriter.writeKeyToFile(primaryKeySavePath, encryptedPrimaryKey, config)
    }
  }

  def retrieveDataKey(primaryKeyPath: String, dataKeySavePath: String,
                      config: Configuration = null): Unit = {
    timing("SimpleKeyManagementService retrieveDataKey") {
      Log4Error.invalidInputError(enrollMap.keySet.contains(_appId) &&
        enrollMap(_appId) == _apiKey, "appid and apikey do not match!")
      Log4Error.invalidInputError(primaryKeyPath != null && primaryKeyPath != "",
        "primaryKeyPath should be specified")
      Log4Error.invalidInputError(dataKeySavePath != null && dataKeySavePath != "",
        "dataKeySavePath should be specified")
      val primaryKeyPlaintext = keyReaderWriter.readKeyFromFile(primaryKeyPath, config)
      Log4Error.invalidInputError(primaryKeyPlaintext.substring(0, 12) == _appId,
        "appid and primarykey should be matched!")
      val randVect = (1 to 16).map { x => Random.nextInt(10) }
      val dataKeyPlaintext = randVect.mkString
      var dataKeyCiphertext = ""
      for(i <- 0 until 16) {
        dataKeyCiphertext += '0' + ((primaryKeyPlaintext(i) - '0') +
          (dataKeyPlaintext(i) - '0')) % 10
      }
      keyReaderWriter.writeKeyToFile(dataKeySavePath, dataKeyCiphertext, config)
    }
  }

  def retrieveDataKeyPlainText(primaryKeyPath: String, dataKeyPath: String,
                               config: Configuration = null): String = {
    timing("SimpleKeyManagementService retrieveDataKeyPlaintext") {
      Log4Error.invalidInputError(enrollMap.keySet.contains(_appId) &&
        enrollMap(_appId) == _apiKey, "appid and apikey do not match!")
      Log4Error.invalidInputError(primaryKeyPath != null && primaryKeyPath != "",
        "primaryKeyPath should be specified")
      Log4Error.invalidInputError(dataKeyPath != null && dataKeyPath != "",
        "dataKeyPath should be specified")
      val primaryKeyCiphertext = keyReaderWriter.readKeyFromFile(primaryKeyPath, config)
      Log4Error.invalidInputError(primaryKeyCiphertext.substring(0, 12) == _appId,
        "appid and primarykey should be matched!")
      val dataKeyCiphertext = keyReaderWriter.readKeyFromFile(dataKeyPath, config)
      var dataKeyPlaintext = ""
      for(i <- 0 until 16) {
        dataKeyPlaintext += '0' + ((dataKeyCiphertext(i) - '0') -
          (primaryKeyCiphertext(i) - '0') + 10) % 10
      }
      dataKeyPlaintext
    }
  }

  private def setAppIdAndKey(appId: String, apiKey: String): Unit = {
    _appId = appId
    _apiKey = apiKey
    enrollMap(_appId) = _apiKey
  }

}

object SimpleKeyManagementService {
  def apply(): SimpleKeyManagementService = {
    val appid = (1 to 12).map(x => Random.nextInt(10)).mkString
    val apikey = (1 to 12).map(x => Random.nextInt(10)).mkString
    new SimpleKeyManagementService(appid, apikey)
  }

  def apply(appID: String, apiKey: String): SimpleKeyManagementService = {
    new SimpleKeyManagementService(appID, apiKey)
  }
}

File: ppml/kms-client/SimpleKeyManagementService.py
import uuid
import base64
from cryptography.fernet import Fernet

enroll_map = {}
_appid = ''
_appkey = ''

def read_encrypted_key_file(encrypted_key_path):
    with open(encrypted_key_path, 'rb') as file:
        original = file.read()
    return original

def write_encrypted_key_file(encrypted_key_path, content):
    with open(encrypted_key_path, 'wb') as file:
        file.write(content)

def set_appid_and_appkey(appid, appkey):
    global _appid
    global _appkey
    _appid = appid
    _appkey = appkey

def enroll():
    while True:
        appid = str(uuid.uuid4().int>>64)[0:12]
        appkey = str(uuid.uuid4().int>>64)[0:12]
        if appid not in enroll_map:
            enroll_map[appid] = appkey
            set_appid_and_appkey(appid, appkey)
            return {appid, appkey}

def retrieve_primary_key(primary_key_path):
    if primary_key_path == '':
        print('[Error] primary_key_save_path should be specified')
        return
    if enroll_map[_appid] != _appkey:
        print('[Error] appid and appkey do not match!')
        return
    encrypted_primary_key = Fernet.generate_key()
    write_encrypted_key_file(primary_key_path, encrypted_primary_key)

def retrieve_data_key(primary_key_path, data_key_save_path):
    if primary_key_path == '':
        print('[Error] primary_key_path should be specified')
        return
    if  data_key_save_path == '':
        print('[Error] data_key_save_path should be specified')
        return
    if enroll_map[_appid] != _appkey:
        print('[Error] appid and appkey do not match!')
        return
    primary_key_plaintext = read_encrypted_key_file(primary_key_path)
    data_key_plaintext = Fernet.generate_key()
    fernet = Fernet(primary_key_plaintext)
    data_key_ciphertext = fernet.encrypt(data_key_plaintext)
    write_encrypted_key_file(data_key_save_path, data_key_ciphertext)

def retrieve_data_key_plaintext(primary_key_path, data_key_path):
    if primary_key_path == '':
        print('[Error] primary_key_path should be specified')
        return
    if  data_key_path == '':
        print('[Error] data_key_path should be specified')
        return
    if enroll_map[_appid] != _appkey:
        print('[Error] appid and appkey do not match!')
        return
    primary_key_ciphertext = read_encrypted_key_file(primary_key_path)
    data_key_ciphertext = read_encrypted_key_file(data_key_path)
    fernet = Fernet(primary_key_ciphertext)
    data_key_plaintext = fernet.decrypt(data_key_ciphertext)
    return data_key_plaintext

